presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java

```diff
@@ -0,0 +1,193 @@
+        this.maxEntriesBeforeSpill = maxEntriesBeforeSpill;
+        this.memorySizeBeforeSpill = memoryLimitBeforeSpill.toBytes();
+        this.spillerFactory = spillerFactory;
+        this.systemMemoryContext = operatorContext.getSystemMemoryContext().newLocalMemoryContext();
+
+        rebuildHashAggregationBuilder();
+    }
+
+    public void processPage(Page page)
+    {
+        checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+        hashAggregationBuilder.processPage(page);
+    }
+
+    public boolean checkFullAndUpdateMemory()
+    {
+        if (isBusy()) {
+            return true;
+        }
+
+        long memorySize = hashAggregationBuilder.getSizeInMemory();
+
+        systemMemoryContext.setBytes(memorySize);
+
+        if (shouldSpill(memorySize)) {
+            spillToDisk();
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isBusy()
+    {
+        return !spillInProgress.isDone();
+    }
+
+    private boolean shouldSpill(long memorySize)
+    {
+        return (maxEntriesBeforeSpill > 0 && hashAggregationBuilder.getGroupCount() > maxEntriesBeforeSpill)
+                || (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+    }
+
+    public Iterator<Page> buildResult()
+    {
+        checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+        if (!spiller.isPresent()) {
+            return hashAggregationBuilder.buildResult();
+        }
+
+        try {
+            // TODO: don't spill here to disk, instead merge disk content with memory content
+            spillToDisk().get();
+            return mergeFromDisk();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        if (spiller.isPresent()) {
+            spiller.get().close();
+            spiller = Optional.empty();
+        }
+    }
+
+    private CompletableFuture<?> spillToDisk()
+    {
+        hashAggregationBuilder.setOutputPartial();
+
+        if (!spiller.isPresent()) {
+            spiller = Optional.of(spillerFactory.get().create(hashAggregationBuilder.buildTypes()));
+        }
+
+        // start spilling process with current content of the hashAggregationBuilder builder...
+        spillInProgress = spiller.get().spill(hashAggregationBuilder.buildResultSorted());
+        // ... and immediately create new hashAggregationBuilder so effectively memory ownership
+        // over hashAggregationBuilder is transferred from this thread to a spilling thread
+        rebuildHashAggregationBuilder();
+
+        return spillInProgress;
+    }
+
+    private Iterator<Page> mergeFromDisk()
+    {
+        checkState(spiller.isPresent());
+
+        MergingHashAggregationBuilder mergingHashAggregationBuilder = new MergingHashAggregationBuilder(
```

Comment:
again it would be cleaner if both `MergingHashAggregationBuilder` and `InMemoryHashAggregationBuilder` are passed as constructor parameters (as `HashAggregationBuilders`). You would need `rebuild()` method in `HashAggregationBuilder`, maybe with some parameters.

Is that feasible with all those arguments? Static implementation binding is not a good code practice.